### PR TITLE
[App Code Check] add check for version and mandatory fields

### PR DESCRIPTION
**Pull request by MorrisJobke**
- ref #17598
- including unit tests for mandatory fields/versions

I nearly forgot this branch and I now added some unit tests. This adds checks for the info.xml and gives some hints that `<version>` in `info.xml` should be prefered to `appinfo/version`.

cc @DeepDiver1975 @karlitschek @PVince81 @nickvergessen

### CODE REVIEW

File: core/command/app/checkcode.php [hunk 1]
```diff
@@ -84,7 +95,7 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 				$output->writeln("<info>Analysing {$filename}</info>");
 			}
 
-			// show error count if there are errros present or the verbosity is high
+			// show error count if there are errrors present or the verbosity is high
```
Comment (nickvergessen) [inline core/command/app/checkcode.php]: you can remove one of the earlier `r`s now, errors doesnt need 4 of them

Comment (MorrisJobke) [inline core/command/app/checkcode.php]: Oh ...didn't noticed them :D

File: core/command/app/checkcode.php [hunk 2]
```diff
@@ -98,6 +109,46 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 			}
 		});
 		$errors = $codeChecker->analyse($appId);
+
+		$infoChecker = new InfoChecker($this->infoParser);
```
Comment (nickvergessen) [inline core/command/app/checkcode.php]: please make this a checker with the parameters?

Comment (nickvergessen) [inline core/command/app/checkcode.php]:
```php
 			->addOption(
 				'checker',
 				'c',
 				InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY,
 				'enable the specified checker(s)',
 				[ 'private', 'deprecation', 'strong-comparison' ]
 			);
```

Comment (MorrisJobke) [inline core/command/app/checkcode.php]: It's not really a code checker :( Therefore it doesn't hook into the `$checkList` that is used for the checker for loop.

Comment (Xenopathic) [inline core/command/app/checkcode.php]: At the very least, it needs to be toggleable. I see what you mean about it not being a code checker, so putting it in `--checker` would be bad, but perhaps something like `--validate-info=true/false` ?

Comment (MorrisJobke) [inline core/command/app/checkcode.php]: Added :)

Comment (MorrisJobke) [inline core/command/app/checkcode.php]: Enabled by default.

**Comment by nickvergessen** [comment]
This falsly breaks shipped apps like the activity app and private apps:
```
./occ app:check-code activity
Mandatory field missing: ocsid
```

ocsid is not mandatory: https://github.com/owncloud/core/blob/master/apps/files/appinfo/info.xml it is optional and "only" strongly suggested for non-shipped apps...

**Comment by MorrisJobke** [comment]
> ocsid is not mandatory: https://github.com/owncloud/core/blob/master/apps/files/appinfo/info.xml it is optional and "only" strongly suggested for non-shipped apps...

Okay. Should I then hardcode the shipped apps? Or should I rely on `<shipped>` (which then can be abused) or what is the way to solve this? Didn't we have already hardcoded list of shipped apps?

**Comment by MorrisJobke** [comment]
> Didn't we have already hardcoded list of shipped apps?

I guess I will take this: `core/shipped.json`

**Comment by nickvergessen** [comment]
Well still then, the field is not mandatory.

**Comment by MorrisJobke** [comment]
> Well still then, the field is not mandatory.

Yep. I will skip the mandatory check for shipped apps ;)

**Comment by nickvergessen** [comment]
> Yep. I will skip the mandatory check for shipped apps ;)

It's not mandatory for any apps:
https://doc.owncloud.org/server/8.1/developer_manual/app/info.html#ocsid

Only id, name, description, licence and author are required

**Comment by MorrisJobke** [comment]
> Only id, name, description, licence and author are required

Updated.

**Comment by nickvergessen** [comment]
```
Deprecated field available: default_enable => 
```
Would be nice to tell what the replacement is.

---
Can't judge this
```
{"reqId":"XUx8d3o7kkeUJCze43HR","remoteAddr":"","app":"PHP","message":"Array to string conversion at \/home\/nickv\/ownCloud\/master\/core\/core\/command\/app\/checkcode.php#127","level":0,"time":"2015-09-24T08:01:18+00:00","method":"--","url":"--"}

Deprecated field available: remote => Array
```

and that
```
{"reqId":"XUx8d3o7kkeUJCze43HR","remoteAddr":"","app":"PHP","message":"Array to string conversion at \/home\/nickv\/ownCloud\/master\/core\/core\/command\/app\/checkcode.php#127","level":0,"time":"2015-09-24T08:01:18+00:00","method":"--","url":"--"}

Deprecated field available: public => Array
```

**Comment by MorrisJobke** [comment]
> Would be nice to tell what the replacement is.

Nothing. Default enable is now handled via shipped apps, because we won't want apps to specify this.



> Can't judge this

I just took the deprecated stuff from documentation. :see_no_evil:

**Comment by nickvergessen** [comment]
@MorrisJobke I ment `Array to string conversion`

**Comment by MorrisJobke** [comment]
> @MorrisJobke I ment Array to string conversion

:see_no_evil: I will fix it

**Comment by MorrisJobke** [comment]
@nickvergessen Updated :)

**Comment by nickvergessen** [comment]
Some more array failures:

> {"reqId":"RLuOl6ucOhn8dg+pT1BH","remoteAddr":"","app":"PHP","message":"Array to string conversion at \/home\/nickv\/ownCloud\/master\/core\/core\/command\/app\/checkcode.php#155","level":0,"time":"2015-09-24T10:28:43+00:00","method":"--","url":"--"}
Optional field available: types => Array

> {"reqId":"RLuOl6ucOhn8dg+pT1BH","remoteAddr":"","app":"PHP","message":"Array to string conversion at \/home\/nickv\/ownCloud\/master\/core\/core\/command\/app\/checkcode.php#159","level":0,"time":"2015-09-24T10:28:43+00:00","method":"--","url":"--"}
Unused field available: dependencies => Array

> {"reqId":"RLuOl6ucOhn8dg+pT1BH","remoteAddr":"","app":"PHP","message":"Array to string conversion at \/home\/nickv\/ownCloud\/master\/core\/core\/command\/app\/checkcode.php#159","level":0,"time":"2015-09-24T10:28:43+00:00","method":"--","url":"--"}
Unused field available: info => Array

**Comment by MorrisJobke** [comment]
Also fixed this more general.

**Comment by nickvergessen** [comment]
Looks good

**Comment by MorrisJobke** [comment]
> Looks good

Is this a review? Please +1 if you're fine with this ;)

**Comment by nickvergessen** [comment]
:+1:

File: lib/private/app/codechecker/infochecker.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		'ocsid',
+		'repository',
+		'require',
+		'requiremin',
+		'types',
+		'version',
+		'website',
+	];
+	private $deprecatedFields = [
+		'default_enable',
```
Comment (DeepDiver1975) [inline lib/private/app/codechecker/infochecker.php]: no

File: lib/private/app/codechecker/infochecker.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		'info',
+		'public',
```
Comment (DeepDiver1975) [inline lib/private/app/codechecker/infochecker.php]: no

File: lib/private/app/codechecker/infochecker.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		'remote',
```
Comment (DeepDiver1975) [inline lib/private/app/codechecker/infochecker.php]: no

File: lib/private/app/codechecker/infochecker.php (hunk 3 above)
Comment (MorrisJobke) [inline lib/private/app/codechecker/infochecker.php]: I just followed the documentation: https://doc.owncloud.org/server/8.1/developer_manual/app/info.html#deprecated

Comment (DeepDiver1975) [inline lib/private/app/codechecker/infochecker.php]: so we better change the docs as well

**Comment by MorrisJobke** [comment]
@DeepDiver1975 Fixed the fields.

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by BernhardPosselt** [comment]
Thanks @MorrisJobke highly appreciated!

**Pull request merged**